examples/bert/bert_model.py

```diff
@@ -0,0 +1,709 @@
+        layer_output = self._output_dense(inner_output)
+        layer_output = self._hidden_dropout(layer_output)
+
+        if self._norm_first:
+            return source_attention_output + layer_output
+
+        # During mixed precision training, layer norm output is always fp32 for
+        # now. Casts fp32 for the subsequent add.
+        layer_output = tf.cast(layer_output, tf.float32)
+        return self._output_layer_norm(layer_output + attention_output)
+
+
+class PositionEmbedding(keras.layers.Layer):
+    """Creates a positional embedding.
+
+    Example:
+    ```python
+    position_embedding = PositionEmbedding(max_length=100)
+    inputs = keras.Input((100, 32), dtype=tf.float32)
+    outputs = position_embedding(inputs)
+    ```
+
+
+    Args:
+        max_length: The maximum size of the dynamic sequence.
+        initializer: The initializer to use for the embedding weights. Defaults
+            to "glorot_uniform".
+        seq_axis: The axis of the input tensor where we add the embeddings.
+
+    Reference: This layer creates a positional embedding as described in
+    [BERT: Pre-training of Deep Bidirectional Transformers for Language
+    Understanding](https://arxiv.org/abs/1810.04805).
+    """
+
+    def __init__(
+        self, max_length, initializer="glorot_uniform", seq_axis=1, **kwargs
+    ):
+
+        super().__init__(**kwargs)
+        if max_length is None:
+            raise ValueError("`max_length` must be an Integer, not `None`.")
+        self._max_length = max_length
+        self._initializer = keras.initializers.get(initializer)
+        self._seq_axis = seq_axis
+
+    def get_config(self):
+        config = {
+            "max_length": self._max_length,
+            "initializer": keras.initializers.serialize(self._initializer),
+            "seq_axis": self._seq_axis,
+        }
+        base_config = super().get_config()
+        return dict(list(base_config.items()) + list(config.items()))
+
+    def build(self, input_shape):
+        dimension_list = input_shape.as_list()
+        width = dimension_list[-1]
+        weight_sequence_length = self._max_length
+
+        self._position_embeddings = self.add_weight(
+            "embeddings",
+            shape=[weight_sequence_length, width],
+            initializer=self._initializer,
+        )
+
+        super().build(input_shape)
+
+    def call(self, inputs):
+        input_shape = tf.shape(inputs)
+        actual_seq_len = input_shape[self._seq_axis]
+        position_embeddings = self._position_embeddings[:actual_seq_len, :]
+        new_shape = [1 for _ in inputs.get_shape().as_list()]
+        new_shape[self._seq_axis] = actual_seq_len
+        new_shape[-1] = position_embeddings.get_shape().as_list()[-1]
+        position_embeddings = tf.reshape(position_embeddings, new_shape)
+        return tf.broadcast_to(position_embeddings, input_shape)
+
+
+class OnDeviceEmbedding(keras.layers.Layer):
```

Comment:
This layer doesn't actually do device placement. It seems to require quite a bit of knowledge of the underlying implementation in order to be used correctly. If possible, I recommend make it an implementation detail of the regular embedding / positional embedding layer.

Reply:
Yeah, this is strictly for TPU support IIUC. We should remove it entirely and make any changes to core embedding layer to do so. Left a TODO.